include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,243 @@
+
+        explicit
+        impl(char_allocator_type const& a)
+        : base(1), a_(a)
+        {
+        }
+
+        static std::size_t ceil_div(std::size_t n, std::size_t d) {
+            return (n + d - 1) / d;
+        }
+
+        char*
+        allocate(std::size_t n, std::size_t size, std::size_t align) override
+        {
+            switch (align) {
+            case alignof(aligned_1_t):
+                if (size == 1) {
+                    return a_.allocate(n);
+                } else {
+                    return reinterpret_cast<char*>(aligned_1_allocator_type(a_).allocate(n * ceil_div(size, align)));
+                }
+            case alignof(aligned_2_t):
+                return reinterpret_cast<char*>(aligned_2_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            case alignof(aligned_4_t):
+                return reinterpret_cast<char*>(aligned_4_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            case alignof(aligned_8_t):
+                return reinterpret_cast<char*>(aligned_8_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            case alignof(aligned_16_t):
+                return reinterpret_cast<char*>(aligned_16_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            default:
+                return reinterpret_cast<char*>(aligned_max_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            }
+        }
+
+        void
+        deallocate(char* p, std::size_t n, std::size_t size, std::size_t align) {
+            switch (align) {
+            case alignof(aligned_1_t): {
+                if (size == 1) {
+                    a_.deallocate(p, n);
+                } else {
+                    aligned_1_allocator_type(a_).deallocate(reinterpret_cast<aligned_1_t*>(p), n * ceil_div(size, align));
+                }
+            break;
+            }
+            case alignof(aligned_2_t):
+                aligned_2_allocator_type(a_).deallocate(reinterpret_cast<aligned_2_t*>(p), n * ceil_div(size, align));
+                break;
+            case alignof(aligned_4_t):
+                aligned_4_allocator_type(a_).deallocate(reinterpret_cast<aligned_4_t*>(p), n * ceil_div(size, align));
+                break;
+            case alignof(aligned_8_t):
+                aligned_8_allocator_type(a_).deallocate(reinterpret_cast<aligned_8_t*>(p), n * ceil_div(size, align));
+                break;
+            case alignof(aligned_16_t):
+                aligned_16_allocator_type(a_).deallocate(reinterpret_cast<aligned_16_t*>(p), n * ceil_div(size, align));
+                break;
+            default:
+                aligned_max_allocator_type(a_).deallocate(reinterpret_cast<std::max_align_t*>(p), n * ceil_div(size, align));
+                break;
+            }
+        }
+
+        void
+        destroy_self() noexcept override
+        {
+            deallocate(reinterpret_cast<char*>(this), 1, sizeof(impl), alignof(impl));
```

Comment:
This is confusing to me, because we have the correctly typed allocator. The original code used the allocator bound to `impl`, why did we change it?

Reply:
Because of all the discussion we had on alignment. The expected type for most allocations is `char`, so we should use `char` to make `allocate` cheaper. Peter's idea.

Reply:
`impl` was already allocated with a correctly typed allocator (or it should have been)

Reply:
This is because the allocator in `base` was bound to `impl` but it's now bound to `char` because of the discussion we've had. In any case, we do have access to `impl` these so I'll rebind it to `impl` then.